File: Turtle.cpp
#include "CONSTANTS.h"
#include "Turtle.h"

#include "World.h"

Turtle::Turtle(World &world, const Point &position) : Animal(world, position, TURTLE_SYMBOL, TURTLE_STRENGTH, TURTLE_INITIATIVE, TURTLE_COLOR, TURTLE_NAME) {}

Organism* Turtle::Clone(const Point &position) {
  return new Turtle(this->world, position);
}

void Turtle::Action() {
  if(CanAction()){
    if (rand() % 4 == 0) {
      Animal::Action();
    }
  }
}

bool Turtle::Collision(Organism *attackerOrganism) {
  if(this->GetSymbol() == attackerOrganism->GetSymbol()){
    this->Breed(attackerOrganism);
    return true;
  } 
  if(attackerOrganism->GetStrength() < 5) {
    this->world.messageManager.AddAttackRepellMessage(attackerOrganism->GetName(), this->GetName());
    return true;
  } else {
    return Animal::Collision(attackerOrganism);
  }
}

File: Java/Turtle.java
import java.util.Random;

public class Turtle extends Animal {
  public Turtle(World world, Point position) {
    super(world, position, Constants.TURTLE_SYMBOL, Constants.TURTLE_STRENGTH, Constants.TURTLE_INITIATIVE, Constants.TURTLE_COLOR, Constants.TURTLE_NAME, Constants.ANIMAL_RANGE);
  }

  @Override
  public final Organism Clone(Point position) {
    return new Turtle(this.world, position);
  }

  @Override
  public final boolean Collision(Organism other) {
    if (other.GetStrength() < 5) {
      this.world.messageManager.AddAttackRepelledMessage(other.GetName(), this.GetName());
      return true;
    }
    return super.Collision(other);
  }

  @Override
  public final void Action() {
    Random random = new Random();
    if (random.nextInt(4) == 0) {
      super.Action();
    }
  }
}
